kernel/ptrace.c

```diff
@@ -0,0 +1,243 @@
+        }
+    }
+    case PTRACE_SINGLESTEP_: {
+        STRACE("ptrace(PTRACE_SINGLESTEP, %d, %#x, %#x)", pid, addr, data);
+        struct task *child = find_child(current, pid);
+        if (child) {
+            child->cpu.tf = true;
+            child->ptrace.stopped = false;
+            notify(&child->ptrace.stop_cond);
+            unlock(&child->ptrace.stop_lock);
+            return 0;
+        } else {
+            return _EPERM;
+        }
+    }
+    case PTRACE_GETREGS_: {
+        STRACE("ptrace(PTRACE_GETREGS, %d, %#x, %#x)", pid, addr, data);
+        struct task *child = find_child(current, pid);
+        if (child) {
+            struct user_regs_struct_ user_regs_ = {};
+            user_regs_.ebx = child->cpu.ebx;
+            user_regs_.ecx = child->cpu.ecx;
+            user_regs_.edx = child->cpu.edx;
+            user_regs_.esi = child->cpu.esi;
+            user_regs_.edi = child->cpu.edi;
+            user_regs_.ebp = child->cpu.ebp;
+            user_regs_.eax = child->cpu.eax;
+//          user_regs_.xds = child->cpu.xds;
+//          user_regs_.xes = child->cpu.xes;
+//          user_regs_.xfs = child->cpu.xfs;
+//          user_regs_.xgs = child->cpu.xgs;
+            user_regs_.orig_eax = child->cpu.eax;
+            user_regs_.eip = child->cpu.eip;
+//          user_regs_.xcs = child->cpu.xcs;
+            user_regs_.eflags = child->cpu.eflags;
+            user_regs_.esp = child->cpu.esp;
+//          user_regs_.xss = child->cpu.xss;
+            if (user_put(data, user_regs_)) {
+                unlock(&child->ptrace.stop_lock);
+                return _EFAULT;
+            }
+            unlock(&child->ptrace.stop_lock);
+        } else {
+            return _EPERM;
+        }
+        return 0;
+    }
+    case PTRACE_SETREGS_: {
+        STRACE("ptrace(PTRACE_SETREGS, %d, %#x, %#x)", pid, addr, data);
+        struct task *child = find_child(current, pid);
+        if (child) {
+            struct user_regs_struct_ user_regs_;
+            if (user_get(data, user_regs_)) {
+                return _EFAULT;
+            } else {
+                child->cpu.ebx = user_regs_.ebx;
+                child->cpu.ecx = user_regs_.ecx;
+                child->cpu.edx = user_regs_.edx;
+                child->cpu.esi = user_regs_.esi;
+                child->cpu.edi = user_regs_.edi;
+                child->cpu.ebp = user_regs_.ebp;
+                child->cpu.eax = user_regs_.eax;
+                // child->cpu.xds = user_regs_.xds;
+                // child->cpu.xes = user_regs_.xes;
+                // child->cpu.xfs = user_regs_.xfs;
+                // child->cpu.xgs = user_regs_.xgs;
+                // child->cpu.eax = user_regs_.orig_eax;
+                child->cpu.eip = user_regs_.eip;
+                // child->cpu.xcs = user_regs_.xcs;
+                child->cpu.eflags = user_regs_.eflags;
+                child->cpu.esp = user_regs_.esp;
+                // child->cpu.xss = user_regs_.xss;
```

Comment:
This is duplicated in POKEUSER, pull into a function? Same with PEEKUSER and GETREGS

Reply:
probably a good idea